src/v2/Utils/Hooks/useFadeTransition.tsx

```diff
@@ -0,0 +1,81 @@
+import { compact } from "lodash"
+import { createRef, RefObject, useCallback, useEffect, useRef } from "react"
+import { wait } from "../wait"
+import { useMode } from "./useMode"
+
+export interface UseFadeTransition {
+  /** Duration of transition in milliseconds */
+  duration?: number
+  /**
+   * If set to `"Out"` all registered elements will
+   * initially have their opacities set to 0
+   * */
+  initialStatus?: "In" | "Out"
+}
+
+export const useFadeTransition = ({
+  duration = 500,
+  initialStatus = "In",
+}: UseFadeTransition = {}) => {
+  const [status, setStatus] = useMode<"In" | "Out">(initialStatus)
+  const [mode, setMode] = useMode<"Resting" | "Transitioning">("Resting")
+
+  // If the `initialStatus` is `"Out"` make sure els are initially hidden
+  useEffect(() => {
+    if (status === "In") return
+
+    const els = compact(refs.current.map(ref => ref.current))
+
+    els.forEach(el => {
+      el.style.opacity = "0"
+    })
+  }, [status])
+
+  const refs = useRef<RefObject<HTMLElement>[]>([])
+
+  /**
+   * @param index The index of the element to register.
+   * Transitions will execute in the order specified by the indexes.
+   * @returns A ref to the element.
+   */
+  const register = useCallback((index: number) => {
```

Comment:
`register` very simply creates and returns a ref; then keeps track of that ref in the index you specify when registering it.